### ISSUE TITLE
feat(finetune): add RLVR — RL on Verifiable Rewards

### ISSUE THREAD

**Issue by noahgift** [paiml/aprender#450]
## Summary

Add Reinforcement Learning on Verifiable Rewards (RLVR) as a fine-tuning
method. RLVR is the final stage of the 4-layer adaptation pipeline — it
uses verifiable reward signals (test pass/fail, math verification, format
checks) to further refine the model.

## CLI

```bash
apr finetune model.apr --method rlvr \
  --reward-fn code-test \
  --data problems.jsonl \
  --rlvr-epochs 5 \
  -o rl-model.apr
```

## Reward Functions

- `code-test`: Execute generated code, reward = test pass rate
- `math-verify`: Verify mathematical answer correctness
- `format-check`: Validate output format compliance (JSON, XML, etc.)
- `custom`: User-provided reward script (reads stdin, writes score to stdout)

## Parameters

- `--reward-fn`: Reward function type
- `--reward-script`: Path to custom reward script (for `--reward-fn custom`)
- `--rlvr-kl-coeff`: KL penalty coefficient (default: 0.05)
- `--rlvr-clip-range`: PPO clip range (default: 0.2)
- `--rlvr-epochs`: RL training epochs (default: 5)

## Data Format (JSONL)

```json
{"prompt": "Write a function that...", "test_cases": ["assert f(1)==1", "assert f(5)==120"]}
```

## Implementation

1. Add `rlvr` variant to FinetuneMethod enum
2. Reward function registry (built-in + custom script)
3. PPO training loop with KL penalty against reference model
4. Code execution sandbox for `code-test` reward (subprocess with timeout)
5. Math verification engine for `math-verify` reward
6. Support LoRA for parameter-efficient RL

## Specification

See `docs/specifications/components/finetune.md` §5.3.

### Contract: `rlvr-training-v1.yaml`

```yaml
metadata:
  description: "RL on Verifiable Rewards — reinforcement learning with verifiable reward signals"
  references:
    - "Schulman et al. (2017) PPO"
    - "Lambert et al. (2024) RLVR"
  depends_on:
    - "adamw-kernel-v1"
    - "lora-algebra-v1"

equations:
  ppo_objective:
    formula: "L_PPO = -E[min(r_t * A_t, clip(r_t, 1-ε, 1+ε) * A_t)]"
    domain: "r_t = π(a|s)/π_old(a|s), ε ∈ (0, 1)"
    invariants:
      - "Clipped ratio bounds policy update magnitude"
      - "Advantage A_t centered: E[A_t] ≈ 0"

  kl_penalty:
    formula: "L_total = L_PPO - λ * KL(π || π_ref)"
    invariants:
      - "λ >= 0 (KL penalty is non-negative weight)"
      - "KL(π || π_ref) >= 0 (KL divergence non-negative)"
      - "Policy doesn't diverge too far from reference"

  verifiable_reward:
    formula: "R(y, x) = verify(y, test_cases(x)) ∈ {0, 1}"
    invariants:
      - "Reward is deterministic for same (output, test_cases)"
      - "Reward ∈ [0, 1] (bounded)"
      - "Verification has timeout (no infinite loops)"

proof_obligations:
  - type: bound
    property: "Clipped ratio bounds"
    formal: "clip(r, 1-ε, 1+ε) ∈ [1-ε, 1+ε]"
  - type: bound
    property: "KL non-negativity"
    formal: "KL(π || π_ref) >= 0"
  - type: invariant
    property: "Reward determinism"
    formal: "verify(y, tests) returns same value for same inputs"
  - type: bound
    property: "Reward boundedness"
    formal: "0 <= R(y, x) <= 1"
  - type: invariant
    property: "Verification timeout"
    formal: "verify() completes within timeout_ms"

falsification_tests:
  - id: FALSIFY-RLVR-001
    rule: "Clip bounds"
    prediction: "clipped ratio always in [1-ε, 1+ε]"
    if_fails: "Clip function implementation incorrect"
  - id: FALSIFY-RLVR-002
    rule: "Reward determinism"
    prediction: "same code + same tests → same reward"
    if_fails: "Non-determinism in code execution sandbox"
  - id: FALSIFY-RLVR-003
    rule: "Timeout enforcement"
    prediction: "infinite loop code returns reward=0, not hang"
    if_fails: "Sandbox timeout not enforced"
  - id: FALSIFY-RLVR-004
    rule: "KL non-negative"
    prediction: "KL(π || π_ref) >= 0 for random policy pairs"
    if_fails: "KL computation bug"
```

### Binding Requirements

Add to `contracts/entrenar/binding.yaml`:
```yaml
  - contract: rlvr-training-v1.yaml
    equation: ppo_objective
    module_path: "entrenar::finetune::rlvr"
    function: ppo_loss
    status: not_implemented

  - contract: rlvr-training-v1.yaml
    equation: verifiable_reward
    module_path: "entrenar::finetune::rlvr"
    function: verify_reward
    status: not_implemented
```

### Verification Targets
- L2: 4 falsification tests
- L3: Kani harness for clip bounds and KL non-negativity
- L4: Lean proof for KL divergence non-negativity (Gibbs inequality)
